aws-iot-thinggroup/src/main/java/software/amazon/iot/thinggroup/UpdateHandler.java

```diff
@@ -42,88 +49,148 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
-
-                proxyClient.injectCredentialsAndInvokeV2(
-                        updateThingGroupRequest,
-                        proxyClient.client()::updateThingGroup
-                );
-                logger.log(String.format("%s %s has successfully been updated.",
-                        ResourceModel.TYPE_NAME, updateThingGroupRequest.thingGroupName()));
-            }
-
-            // update the Tags as specified in the resource model by generating a diff of current and desired Tags
-            updateTags(proxyClient, resourceModel, desiredTags);
-        } catch (Exception e) {
-            return Translator.translateExceptionToProgressEvent(resourceModel, e, logger);
+        validatePropertiesAreUpdatable(newResourceModel, prevResourceModel);
+
+        if (isDynamicThingGroup(checkForThingGroup(newResourceModel.getThingGroupName(), proxyClient, OPERATION))) {
+            return ProgressEvent.progress(newResourceModel, callbackContext)
+                    .then(progress ->
+                            proxy.initiate(CALL_GRAPH, proxyClient, newResourceModel, callbackContext)
+                                    .translateToServiceRequest(Translator::translateToFirstDynamicThingGroupUpdateRequest)
+                                    .makeServiceCall(this::updateDynamicThingGroupResource)
+                                    .progress())
+                    .then(progress -> updateResourceTags(proxy, proxyClient, progress, request, newResourceModel))
+                    .then(progress -> ProgressEvent.defaultSuccessHandler(newResourceModel));
+        } else {
+            return ProgressEvent.progress(newResourceModel, callbackContext)
+                    .then(progress ->
+                            proxy.initiate(CALL_GRAPH, proxyClient, newResourceModel, callbackContext)
+                                    .translateToServiceRequest(Translator::translateToUpdateThingGroupRequest)
+                                    .makeServiceCall(this::updateThingGroupResource)
+                                    .progress())
+                    .then(progress -> updateResourceTags(proxy, proxyClient, progress, request, newResourceModel))
+                    .then(progress -> ProgressEvent.defaultSuccessHandler(newResourceModel));
+        }
+    }
+
+    private void validatePropertiesAreUpdatable(ResourceModel newResourceModel, ResourceModel prevResourceModel) {
+        if (!StringUtils.equals(newResourceModel.getThingGroupName(), prevResourceModel.getThingGroupName())) {
+            throwCfnNotUpdatableException("ThingGroupName");
+        } else if (StringUtils.isNotEmpty(newResourceModel.getArn()) &&
+                !StringUtils.equals(newResourceModel.getArn(), prevResourceModel.getArn())) {
+            throwCfnNotUpdatableException("Arn");
         }
 
-        return ProgressEvent.defaultSuccessHandler(resourceModel);
+        // check the case for switching between dynamic to static thing group and vice-versa
```

Comment:
this is interesting, afaik we dont allow that through cli right? is this something that cfn allows?